File: SampleFRIT/SampleFRIT.py
import numpy as np
from scipy.signal import lti, dlsim, dlti
import csv

# Simulation of FIRT

# Initial settings
Ts = 0.01
A = np.array([[1, 1], [0, -2]])
B = np.array([[0], [1]])
x0 = np.array([0, 0])
n = A.shape[0]
F_ini = np.array([-0.8, 2.0])
Hd = [
    dlti([1], [1, -0.5, 0], dt=Ts),
    dlti([1, -1], [1, -0.5, 0], dt=Ts)
]

# Simulation
k = np.arange(0, 51)  # Step count
N = len(k)
t = np.arange(0, N*Ts, Ts)  # Time vector

# Rectangular wave signal
v = np.zeros(N)
v[2:7] = 1

# State update
x = np.zeros((N, 2))
x[0, :] = x0
for index in range(1, N):
    u = np.dot(F_ini, x[index-1, :]) + v[index-1]
    temp = np.dot(A, x[index-1, :]) + B.flatten() * u
    x[index, :] = temp

# Initial data
x_ini = x
u_ini = np.dot(F_ini, x_ini.T) + v

# Minimize objective function
Gamma = np.zeros(n * N)
W = np.zeros((n * N, n))
for j in range(n):
    # Using lsim for discrete-time systems
    _, y_out1, _ = dlsim(Hd[j], u_ini, t)
    Gamma[j * N:(j + 1) * N] = x_ini[:, j] - y_out1.flatten()

    _, y_out2, _ = dlsim(Hd[j], x_ini[:, 0], t)
    _, y_out3, _ = dlsim(Hd[j], x_ini[:, 1], t)
    W[j * N:(j + 1) * N, :] = np.column_stack((y_out2.flatten(), y_out3.flatten()))

# Write Gamma, W, W'*W to .csv files
np.savetxt('Gamma.csv', Gamma, delimiter=',')
np.savetxt('W.csv', W, delimiter=',')
np.savetxt('Psi.csv', np.dot(W.T, W), delimiter=',')

# Calculate optimal state feedback gain
F_ast = -np.dot(np.dot(Gamma.T, W), np.linalg.inv(np.dot(W.T, W)))

# Display results
print('F_ast =')
print(F_ast)

File: SampleFRIT/SampleFRIT.m
% FIRTのシミュレーション
clear;clc;

% 初期設定
Ts = 0.01;
A = [1 1; 0 -2];
B = [0; 1];
x0 = [0 0];
n = size(A, 1);
F_ini = [-0.8 2.0];
Hd = [
    tf(1, [1 -0.5 0], Ts);
    tf([1 -1], [1 -0.5 0], Ts)
];

% シミュレーション
k = 0:1:50; % ステップ数
N = length(k);
t = 0:Ts:(N-1)*Ts; % 時間ベクトル

% 短形波信号
v = zeros(1,N);
for index = 1:N
    if 1<index && index<7
        v(index) = 1;
    else
        v(index) = 0;
    end
end

% 状態の更新
x = zeros(N, 2);
x(1,:) = x0;
for index = 2:N
    u = F_ini * x(index-1,:)' + v(index-1);
    temp = A * x(index-1,:)' + B * u;
    x(index,:) = temp';
end

% 初期データ
x_ini = x;
u_ini = F_ini * x_ini' + v;

% 評価関数の最小化
Gamma = zeros(n*N, 1);
W = zeros(n*N, n);
for j = 1:n
    Gamma((j-1)*N+1:j*N) = x_ini(:, j) - lsim(Hd(j), u_ini, t);
    W((j-1)*N+1:j*N, :) = [lsim(Hd(j), x_ini(:, 1)', t) lsim(Hd(j), x_ini(:, 2)', t)];
end

% Gamma, W, W'*Wを.csvファイルに書き出す
csvwrite('Gamma.csv', Gamma);
csvwrite('W.csv', W);
csvwrite('Psi.csv', W'*W);

% 最適な状態フィードバックゲインの計算
F_ast = -Gamma' * W * inv(W' * W);

% 結果の表示
disp('F_ast = ');
disp(F_ast);
